### PR TITLE
Submodules for user directory types

### PR DESCRIPTION
**Pull request by alexander-yevsyukov**
This PR changes the architecture of the library. 

The core sub-domain is based on events and Projection state types like `User`, `UserAccount`, and `Group`. It's what the `users-client` artefact provides. Correspondingly, the `users-server` artefact provides the server-side implementation for `Projection` types that build `User`, `Group`, etc.

User-management implementations, like `users-db`, are responsible for producing the events that `users-server` consumes for building the projections. 

The `users-db` module added in this PR serves as an example for now it can work. This PR does not provide full implementation of the user management based on locally-managed database with users, group, and all.

The code related to authentication, organization, org. units, and roles is extracted into separate repositories. It is expected that it would be developed separately. It may turn that, authentication (at least partially) would be a part of Users library. But for now it was extracted for the sake of simplicity.

Other changes:
  * Gradle files were ported to Kotlin Script.
  * Some commands, events, and entity states were renamed to improve the language.

### CODE REVIEW

**Comment by alexander-yevsyukov** [comment]
@armiol, this is the PR with the draft for new `Client` and `Session` API.

**Comment by alexander-yevsyukov** [comment]
@armiol, @dmdashenkov PTAL. Significant part of this PR is deleted or moved into other repositories. So it's not real 208 files.

File: build.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+                    "io.spine:spine-time:$spineTimeVersion"
+            )
+        }
+    }
+}
+
+plugins {
+    java
+    idea
+    id("com.google.protobuf").version(io.spine.gradle.internal.Deps.versions.protobufPlugin)
```
Comment (dmdashenkov) [inline build.gradle.kts:50]: Please suppress the warning. Otherwise, IDEA will replace this FQN with an import and break the build.

Comment (dmdashenkov) [reply build.gradle.kts:50]: Same below and in the `buildscript` section.

File: build.gradle.kts [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+        targetCompatibility = JavaVersion.VERSION_1_8
+    }
+
+    DependencyResolution.defaultRepositories(repositories)
+
+    dependencies {
+        errorprone(Deps.build.errorProneCore)
+        errorproneJavac(Deps.build.errorProneJavac)
+        // For dependencies config. based on version of Java, see:
+        //  https://github.com/epeee/junit-jupiter-extension-testing/blob/57b7ba75ab64ed8c229d2a5b14a954d6ae359189/gradle/errorprone.gradle
```
Comment (dmdashenkov) [inline build.gradle.kts:115]: I don't think this note is necessary. It should be explained in `deps.kt`.

File: build.gradle.kts [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+                        "io.spine:spine-base:$spineBaseVersion",
+                        "io.spine:spine-base:$spineCoreVersion",
+                        "io.spine:spine-time:$spineTimeVersion"
+                )
+            }
+        }
+    }
+    DependencyResolution.excludeProtobufLite(configurations)
+
+    /* Uncomment this block if you need to display console output during the Gradle build.*/
```
Comment (dmdashenkov) [inline build.gradle.kts:142]: Please remove this comment.

File: users-client/src/main/java/io/spine/users/event/AccountAware.java [hunk 4]
```diff
@@ -18,24 +18,25 @@
...
 
 /**
- * A factory for creating {@link Directory} instances.
+ * The interface common for events referring a user account.
  */
-@FunctionalInterface
-public interface DirectoryFactory {
+@Immutable
+@GeneratedMixin
+public interface AccountAware extends EventMessage {
```
Comment (dmdashenkov) [inline users-client/src/main/java/io/spine/users/event/AccountAware.java:33]: Its probably `AccountAwareEvent`, since it extends `EventMessage`.

File: users-client/src/main/java/io/spine/users/event/GroupAware.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+import io.spine.annotation.GeneratedMixin;
+import io.spine.base.EventMessage;
+import io.spine.users.GroupId;
+
+/**
+ * Common interface for events referring a group.
+ */
+@Immutable
+@GeneratedMixin
+public interface GroupAware extends EventMessage {
```
Comment (dmdashenkov) [inline users-client/src/main/java/io/spine/users/event/GroupAware.java:33]: Same as above.

File: users-db/src/main/java/io/spine/users/server/db/GroupAccountPart.java [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+        Group.Builder builder = builder();
+        builder.setId(e.getGroup())
+               .setDisplayName(e.getDisplayName())
+               .setDescription(e.getDescription())
+               .setEmail(e.getEmail());
+    }
+
+    @Apply
+    private void on(
+            @SuppressWarnings({
```
Comment (dmdashenkov) [inline users-db/src/main/java/io/spine/users/server/db/GroupAccountPart.java:131]: Please move the suppressions to the method level. I know its a bit of an overkill, but this way we don't lose the fluency of an applier method declaration.

File: users-db/src/main/java/io/spine/users/server/db/GroupMembersPart.java [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+        Optional<Member> found =
+                members().filter(m -> m.getKindCase() == GROUP)
+                         .filter(m -> group.equals(m.getGroup()))
+                         .findFirst();
+        return found;
+    }
+
+    private Stream<Member> members() {
+        return state().getMemberList()
+               .stream();
```
Comment (dmdashenkov) [inline users-db/src/main/java/io/spine/users/server/db/GroupMembersPart.java:157]: Please fix the alignment.

File: version.gradle.kts [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+project.extra.apply {
+    this["versionToPublish"] = "0.0.3"
+    this["spineCoreVersion"] = "1.6.0"
+    this["spineBaseVersion"] = "1.6.0"
```
Comment (dmdashenkov) [inline version.gradle.kts:24]: Please use the latest version of base.

**Review by dmdashenkov — CHANGES_REQUESTED** [review]
It seems strange to me that there are 2 entities called `UserAccount`, an aggregate part and a projection. I realize that one is part of a DB-specific model while the other is more general. Still would be nice to be able to tell one from the other without checking the package name. Also, I doubt that we would call it the same name in a real-life conversation.

On a more general note, I wonder if projections are the way to go here. Are we emphasizing that the source of truth is "out there and not here"? If so, will we be able to avoid duplicating logic of building this model across different implementations?

File: users-client/src/main/proto/spine/users/events.proto [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+    string new_description = 2 [(required) = true];
+
+    // An old human-friendly description of the group.
+    string old_description = 3 [(required) = true];
+}
+
+// A group email address was changed.
+message GroupEmailChanged {
+    option (is).java_type = "GroupAware";
+    option (required_field) = "new_email|old_email";
```
Comment (armiol) [inline users-client/src/main/proto/spine/users/events.proto:140]: I thought we needed the `new_email` to be required and not equal to `old_email`.

The current definition says that a group may have no email — in case someone has erased it, `new_email` is going to be empty and `old_email` is going to be filled in.

File: users-client/src/main/proto/spine/users/events.proto [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+    core.UserId user = 1 [(required) = true];
+
+    // An ID of the group.
+    users.GroupId group = 2 [(required) = true];
+
+    // A role of a user in the group.
+    Role role = 3 [(required) = true];
+}
+
+// A user was removed a group.
```
Comment (armiol) [inline users-client/src/main/proto/spine/users/events.proto:166]: "removed from a group".

File: users-client/src/main/proto/spine/users/events.proto [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+    option (is).java_type = "GroupAware";
+
+    // The group was removed from the parent group.
+    GroupId group = 1 [(required) = true];
+
+    // The parent group.
+    GroupId parent_group = 2 [(required) = true];
+
+    // The role which the group had in the parent group, if available.
+    Role role = 3 [(required) = true];
```
Comment (armiol) [inline users-client/src/main/proto/spine/users/events.proto:207]: The description says "if available". So I assume this field isn't required.

File: users-client/src/main/proto/spine/users/identifiers.proto [hunk 12]
```diff
@@ -20,28 +20,27 @@
...
-import "spine/users/identifiers.proto";
+// The ID of a user group.
+message GroupId {
 
-// A new role was created.
-//
-message RoleCreated {
-
-    RoleId id = 1;
+    string value = 1;
```
Comment (armiol) [inline users-client/src/main/proto/spine/users/identifiers.proto:36]: I assume we want to leave this field without any restrictions on its content, as we don't know which ID structure third parties — that we integrate with — have. Is it so?

File: users-client/src/main/proto/spine/users/membership.proto [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+message Membership {
+
+    // The group to which the member belongs.
+    GroupId group = 1 [(required) = true];
+
+    // The role in the group.
+    Role role = 2 [(required) = true];
+}
+
+// To which groups this user belongs?
```
Comment (armiol) [inline users-client/src/main/proto/spine/users/membership.proto:81]: While I understand it's for brevity, I would avoid the questions and go with 
`The groups to which this user belongs`.

The same applies to the `GroupMembership` description down below.

File: users-client/src/main/proto/spine/users/membership.proto [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+    // Groups to which the user belongs.
+    //
+    // Empty if the user is not a member of any group.
+    //
+    repeated Membership direct = 2;
+
+    // Groups to which the user belongs via other groups.
+    //
+    // Empty if the the groups the user belongs to do not belong to other groups.
+    // This means that if the `direct` field is empty, this field must be empty too.
```
Comment (armiol) [inline users-client/src/main/proto/spine/users/membership.proto:97]: So, `indirect` memberships include `direct` memberships? If so, I would call this field "all". See below (notice the updated description of the `direct` field as well):

```proto
...
    // Groups to which the user belongs directly.
    //
    // Empty if the user is not a member of any group.
    //
    repeated Membership direct = 2;

    // All groups to which the user belongs.
    //
    // Includes the groups to which the user belongs directly and the groups of groups.
    //
    // Empty if the the groups the user belongs to do not belong to other groups.
    // This means that if the `direct` field is empty, this field must be empty too.
    //
    repeated Membership all = 3;
...
```

**Review by armiol — CHANGES_REQUESTED** [review]
@alexander-yevsyukov 39/208 reviewed. I have completed the `users-client` code part. Continuing onto the rest of the modules.

File: users-db/src/main/java/io/spine/users/server/db/ContextExtensions.java [hunk 15]
```diff
@@ -18,40 +18,33 @@
...
-     */
-    protected @Nullable Directory createDirectory() {
-        return null;
+    /** Prevents instantiation of this utility class. */
+    private ContextExtensions() {
     }
 
-    protected final @Nullable Directory directory() {
-        return directory;
+    public static BoundedContextBuilder withDatabase(BoundedContextBuilder builder) {
```
Comment (armiol) [inline users-db/src/main/java/io/spine/users/server/db/ContextExtensions.java:39]: Please document.

File: users-db/src/main/java/io/spine/users/server/db/GroupMembersPart.java [hunk 16]
```diff
@@ -0,0 +1,166 @@
...
+    @Apply
+    private void on(GroupRemovedFromGroup event) {
+        GroupId group = event.getGroup();
+        Optional<Member> member = findMember(group);
+        member.ifPresent(this::removeMember);
+    }
+
+    private Optional<Member> findMember(UserId user) {
+        Optional<Member> found =
+                members().filter(m -> m.getKindCase() == USER)
```
Comment (armiol) [inline users-db/src/main/java/io/spine/users/server/db/GroupMembersPart.java:141]: It seems we may shorten this to 
```java
private Optional<Member> findMember(UserId user) {
        Optional<Member> found =
                members().filter(m -> user.equals(m.getUser()))
                         .findFirst();
        return found;
    }
```
Given the passed `user` is not default (which should be guaranteed by the event constraints in the first place), we are going to be fine.

File: users-db/src/main/java/io/spine/users/server/db/GroupMembersPart.java [hunk 17]
```diff
@@ -0,0 +1,166 @@
...
+                         .filter(m -> user.equals(m.getUser()))
+                         .findFirst();
+        return found;
+    }
+
+    private Optional<Member> findMember(GroupId group) {
+        Optional<Member> found =
```
Comment (armiol) [inline users-db/src/main/java/io/spine/users/server/db/GroupMembersPart.java:148]: Same as above:
```java
    private Optional<Member> findMember(GroupId group) {
        Optional<Member> found =
                members().filter(m -> group.equals(m.getGroup()))
                         .findFirst();
        return found;
    }
```

File: users-db/src/main/proto/spine/users/db/commands.proto [hunk 18]
```diff
@@ -0,0 +1,167 @@
...
+// until the account is resumed.
+//
+message SuspendUserAccount {
+
+    // The ID of the account to suspend.
+    core.UserId account = 1;
+}
+
+// Activate previously suspended user account.
+message ResumeUserAccount {
```
Comment (armiol) [inline users-db/src/main/proto/spine/users/db/commands.proto:68]: Just in case, see my suggestion on `resume` -> `restore` above.

File: users-db/src/main/proto/spine/users/db/commands.proto [hunk 19]
```diff
@@ -0,0 +1,167 @@
...
+    string display_name = 2 [(required) = true];
+
+    // A human-friendly description of a group.
+    string description = 3;
+
+    // An email address of the group, if applicable.
+    net.EmailAddress email = 4;
+}
+
+// Deletes a group.
```
Comment (armiol) [inline users-db/src/main/proto/spine/users/db/commands.proto:89]: Same as above. I would say `Delete` instead of `Deletes`.

File: users-db/src/main/proto/spine/users/db/rejections.proto [hunk 20]
```diff
@@ -0,0 +1,118 @@
...
+}
+
+// An operation is not available because an account with the passed ID was previously deleted.
+message UnavailableForTerminatedAccount {
+
+    // The ID of the account which existed before.
+    core.UserId account = 1;
+}
+
+// A command requests creation of a group with an ID which already assigned to another group.
```
Comment (armiol) [inline users-db/src/main/proto/spine/users/db/rejections.proto:74]: Let's say `A command requested...`, as in similar cases.

File: users-db/src/main/proto/spine/users/db/rejections.proto [hunk 21]
```diff
@@ -0,0 +1,118 @@
...
+message GroupsCannotFormCycles {
+
+    // The parent group to which addition of the group requested.
+    GroupId parent_group = 1 [(required) = true];
+
+    // The group that closes a cycle.
+    GroupId closing_group = 2 [(required) = true];
+}
+
+// An operation is not available because a group the passed ID was previously deleted.
```
Comment (armiol) [inline users-db/src/main/proto/spine/users/db/rejections.proto:113]: `...a group with the passed ID...`

File: users-db/src/main/proto/spine/users/db/rejections.proto [hunk 22]
```diff
@@ -0,0 +1,118 @@
...
+}
+
+// A command requested removal of a group, but it is not a member of a group removal from which
+// was requested.
+message GroupIsNotMember {
+
+    // The group from which removal of the group was requested.
+    GroupId parent_group = 1 [(required) = true];
+
+    // The group removal of whom from the group was requested.
```
Comment (armiol) [inline users-db/src/main/proto/spine/users/db/rejections.proto:98]: `...of which...`

File: users-db/src/test/java/io/spine/users/server/db/DbCommandTest.java [hunk 23]
```diff
@@ -0,0 +1,37 @@
...
+package io.spine.users.server.db;
+
+import io.spine.base.CommandMessage;
+import io.spine.base.EntityState;
+import io.spine.base.EventMessage;
+import io.spine.server.BoundedContextBuilder;
+import io.spine.server.entity.Entity;
+import io.spine.users.server.CommandTest;
+
+public abstract class DbCommandTest<I, C extends CommandMessage, E extends EventMessage,
```
Comment (armiol) [inline users-db/src/test/java/io/spine/users/server/db/DbCommandTest.java:30]: As it is an abstract base type, please document its purpose.

File: users-db/src/test/java/io/spine/users/server/db/DbExtensionsTest.java [hunk 24]
```diff
@@ -18,14 +18,15 @@
...
+package io.spine.users.server.db;
 
-@ParametersAreNonnullByDefault
-@CheckReturnValue
-package io.spine.users.server.orgunit;
+import io.spine.server.BoundedContextBuilder;
+import io.spine.users.server.UsersContextTest;
 
-import com.google.errorprone.annotations.CheckReturnValue;
+public abstract class DbExtensionsTest extends UsersContextTest {
```
Comment (armiol) [inline users-db/src/test/java/io/spine/users/server/db/DbExtensionsTest.java:26]: Please document this abstract base class.

File: users-db/src/test/java/io/spine/users/server/db/UserAccountTest.java [hunk 25]
```diff
@@ -0,0 +1,130 @@
...
+
+                context.receivesCommand(createUserAccount(user));
+
+                assertEvent(UnavailableForTerminatedAccount
+                                    .newBuilder()
+                                    .setAccount(user)
+                                    .build());
+            }
+        }
+
```
Comment (armiol) [inline users-db/src/test/java/io/spine/users/server/db/UserAccountTest.java:115]: Please kill this empty line.

File: users-db/src/test/java/io/spine/users/server/db/given/Command.java [hunk 26]
```diff
@@ -0,0 +1,148 @@
...
+import static io.spine.users.server.db.given.Given.anotherGroupName;
+import static io.spine.users.server.db.given.Given.groupDescription;
+import static io.spine.users.server.db.given.Given.groupEmail;
+import static io.spine.users.server.db.given.Given.groupName;
+import static io.spine.users.server.db.given.Given.newGroupEmail;
+
+/**
+ * Test commands for DB-based extensions of the Users context.
+ */
+public class Command {
```
Comment (armiol) [inline users-db/src/test/java/io/spine/users/server/db/given/Command.java:50]: This class seems `final`.

File: users-db/src/test/java/io/spine/users/server/db/package-info.java [hunk 27]
```diff
@@ -18,9 +18,9 @@
  * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
 
-@CheckReturnValue
 @ParametersAreNonnullByDefault
-package io.spine.users.server.signin.given;
+@CheckReturnValue
+package io.spine.users.server.db;
```
Comment (armiol) [inline users-db/src/test/java/io/spine/users/server/db/package-info.java:23]: We seem to already have a `package-info` in this package. It is defined in the `main/java` source root.

**Review by armiol — CHANGES_REQUESTED** [review]
@alexander-yevsyukov 64/208 done. Both `users-client` and `users-db` reviewed. Hopping onto the next module now.

File: users-google/src/main/proto/spine/users/google/group.proto [hunk 28]
```diff
@@ -82,15 +82,13 @@ message GoogleGroupLifecycle {
...
 }
 
-// A projection that maps a `GoogleGroupId` to the corresponding `GroupId`.
-//
-// This projection is used to match IDs when mapping external events from Directory API to
-// `Users` bounded context events.
-//
-message GoogleIdMappingView {
+// Maps a Google Group ID as obtained from Google Directory ID to corresponding
+// `GroupId` of the Users context.
```
Comment (armiol) [inline users-google/src/main/proto/spine/users/google/group.proto:86]: The context in this model is named `Google Directory`. So, I would add some text explaining that you mean not this context, but another. Let's discuss further once we meet.

File: users-google/src/test/java/io/spine/users/google/server/given/GoogleGroupTestEnv.java [hunk 29]
```diff
@@ -18,20 +18,18 @@
...
+import io.spine.users.google.server.GoogleGroupLifecyclePm;
 
 import static io.spine.base.Identifier.newUuid;
 
 /**
  * Test environment for testing {@link GoogleGroupLifecyclePm}.
- *
- * @author Vladyslav Lubenskyi
  */
 public class GoogleGroupTestEnv {
```
Comment (armiol) [inline users-google/src/test/java/io/spine/users/google/server/given/GoogleGroupTestEnv.java:34]: I understand the tests are disabled, but for the sake of reviewing, all the classes in this `.given` package should be `final`.

**Review by armiol — CHANGES_REQUESTED** [review]
@alexander-yevsyukov `users-google` also done (88/208 reviewed so far in this PR). Please see my comments.

File: users-client/src/main/java/io/spine/users/event/AccountAware.java (hunk 4 above)
Comment (alexander-yevsyukov) [reply users-client/src/main/java/io/spine/users/event/AccountAware.java:33]: We're in the `event` package. Anything here is most likely an event. Why don't we DRY?

File: users-db/src/main/java/io/spine/users/server/db/GroupAccountPart.java (hunk 6 above)
Comment (alexander-yevsyukov) [reply users-db/src/main/java/io/spine/users/server/db/GroupAccountPart.java:131]: Right!

File: users-client/src/main/proto/spine/users/events.proto (hunk 9 above)
Comment (alexander-yevsyukov) [reply users-client/src/main/proto/spine/users/events.proto:140]: `new_email` != `old_email` is a separate story. I don't think our validation supports it.

Comment (alexander-yevsyukov) [reply users-client/src/main/proto/spine/users/events.proto:140]: I think there can be general cases when a group had an email address and then lose it. Suppose it's not a Google Group, but just a buch of people who used to have a mailing list. And then, they decided to go to Slack so the group no longer needs an email. So, they ditch the email address and the event reflects that the association with group email address is no longer relevant.

Comment (alexander-yevsyukov) [reply users-client/src/main/proto/spine/users/events.proto:140]: I thought about removing the `email` field from the `Group` at all so that the group thing is formulated in more abstract term. But then I leaned toward practicality. Email address is going to be used. At least when it comes to integrating with Google, which is one of the first implementations we're going to have as long as we need it most.

File: users-client/src/main/proto/spine/users/membership.proto (hunk 14 above)
Comment (alexander-yevsyukov) [reply users-client/src/main/proto/spine/users/membership.proto:97]: I've clarified the description. I want to keep direct and indirect separately. It's easy to combine them, if needed.

Just in case Google announced the [support for indirect group membership](https://gsuiteupdates.googleblog.com/2020/10/new-apis-view-visualize-google-group-membership.html).

File: users-client/src/main/proto/spine/users/identifiers.proto (hunk 12 above)
Comment (alexander-yevsyukov) [reply users-client/src/main/proto/spine/users/identifiers.proto:36]: Exactly.

File: users-google/src/main/java/io/spine/users/google/server/IdMapping.java [hunk 30]
```diff
@@ -18,37 +18,29 @@
...
 /**
- * The repository for {@link GoogleIdMappingProjection}.
- *
- * @author Vladyslav Lubenskyi
+ * The repository for {@link Entity}.
  */
-public class GoogleIdMappingRepository extends ProjectionRepository<GoogleIdMappingViewId,
-                                                                    GoogleIdMappingProjection,
-                                                                    GoogleIdMappingView> {
+public class IdMapping
```
[the remark this answers is not in the document]
Comment (alexander-yevsyukov) [reply users-google/src/main/java/io/spine/users/google/server/IdMapping.java:37]: As discussed vocally, let's keep it this way for now. Integration with Google Workspace API may change significantly really soon.

File: users-client/src/main/proto/spine/users/events.proto (hunk 11 above)
Comment (alexander-yevsyukov) [reply users-client/src/main/proto/spine/users/events.proto:207]: Let's have all the `role` field required. It would simplify a bit the code consuming the events.

**Review by armiol — APPROVED** [review]
@alexander-yevsyukov LGTM.

**Review by dmdashenkov — APPROVED** [review]
